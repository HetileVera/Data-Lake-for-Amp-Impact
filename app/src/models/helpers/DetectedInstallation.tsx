import { types, Instance } from "mobx-state-tree";

/**
 * Represents a short information about a detected installation
 */
export const DetectedInstallation = types
  .model("DetectedInstallation", {
    id: "",
    webUrlOrigin: "",
    reachable: true,
    region: "",
    installationJson: types.maybe(types.frozen()),
  })

  .views((self) => ({
    get startDate(): Date | undefined {
      return self.installationJson?.startDate;
    },

    get startedBy(): string | undefined {
      return self.installationJson?.startedBy;
    },

    get accountId(): string | undefined {
      return self.installationJson?.accountId;
    },

    get appFlowConnectionName(): string | undefined {
      return self.installationJson?.appFlowConnectionName;
    },
  }));

export interface IDetectedInstallation extends Instance<typeof DetectedInstallation> {}
