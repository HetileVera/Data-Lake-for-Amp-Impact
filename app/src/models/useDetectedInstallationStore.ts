import { isStoreError, isStoreLoading, isStoreReady, isStoreReLoading } from "models/BaseStore";

import { DetectedInstallationStore } from "./DetectedInstallationsStore";

// Create an instance of this store
const store = DetectedInstallationStore.create({});

export function useDetectedInstallationStore() {
  return {
    isError: isStoreError(store),
    isReady: isStoreReady(store),
    isLoading: isStoreLoading(store),
    isReloading: isStoreReLoading(store),
    store,
  };
}
