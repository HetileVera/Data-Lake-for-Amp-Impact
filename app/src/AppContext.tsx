import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
