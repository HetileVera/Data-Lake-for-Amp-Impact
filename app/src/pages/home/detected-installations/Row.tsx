import React from "react";
import { Box, Flex, Icon, Text } from "@chakra-ui/react";
import { observer } from "mobx-react";
import { BsCheckCircleFill, BsCircle } from "react-icons/bs";

import { awsRegionsMap } from "data/aws-regions";
import { useColorScheme } from "models/useColorScheme";
import { IDetectedInstallation } from "models/helpers/DetectedInstallation";
import { TimeAgo } from "components/TimeAgo";

interface IRow {
  item: IDetectedInstallation;
  selected: boolean;
  onSelected: (id: string) => void;
}

export const Row = observer(({ item, selected, onSelected }: IRow) => {
  const { tone } = useColorScheme();
  let props: any;

  const getRegionLabel = (name: string) => awsRegionsMap[name].label || name;

  if (selected) props = { color: tone(700), borderColor: tone(600), bg: tone(75) };
  else if (item.reachable) props = { cursor: "pointer", borderColor: tone(300), color: tone(600), onClick: () => onSelected(item.id) };
  else props = { color: "gray.300", cursor: "not-allowed", borderColor: "gray.300" };

  return (
    <Flex borderWidth={1} borderColor={tone(100)} borderRadius="md" p={3} mt={4} {...props}>
      {selected && <Icon as={BsCheckCircleFill} color={tone(700)} fontSize="large" ml={3} mr={2.5} mt={1} />}
      {!selected && <Icon as={BsCircle} color={!item.reachable ? "gray.300" : tone(500)} fontSize="large" ml={3} mr={2.5} mt={1} />}
      <Box flex={1} fontSize="sm">
        <Flex justifyContent="space-between">
          <Text fontWeight="semibold">Data Lake - {item.id}</Text>
          <Text fontSize="xs">{getRegionLabel(item.region)}</Text>
        </Flex>
        {item.reachable && (
          <>
            <Box fontSize="xs">
              <TimeAgo time={item.startDate} />
              <Text display="inline-block" ml={2} mr={2}>
                &#8226;
              </Text>
              by {item.startedBy}
            </Box>
            <Box fontSize="xs">Connection name: {item.appFlowConnectionName}</Box>
          </>
        )}
        {!item.reachable && <Box fontSize="xs">Initiated from {item.webUrlOrigin}</Box>}
      </Box>
    </Flex>
  );
});
