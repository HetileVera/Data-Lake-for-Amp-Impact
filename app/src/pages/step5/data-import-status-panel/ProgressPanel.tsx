import React from "react";
import { observer } from "mobx-react";
import { Progress } from "@chakra-ui/react";

import { useImportStatusStore } from "models/ImportStatusStore";
import { isStoreLoading, isStoreReLoading } from "models/BaseStore";

export const ProgressPanel = observer(({ colorScheme }: { colorScheme: string }) => {
  const { isError, store } = useImportStatusStore();

  if (isError) return null;
  if (!isStoreLoading(store) && !isStoreReLoading(store)) return null;

  return <Progress size="sm" isIndeterminate colorScheme={colorScheme} bg={`${colorScheme}.100`} borderRadius="lg" mb={3} />;
});
