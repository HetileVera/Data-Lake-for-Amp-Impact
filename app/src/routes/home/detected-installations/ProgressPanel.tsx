import React from "react";
import { Progress, Text } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useDetectedInstallationStore } from "models/useDetectedInstallationStore";
import { useColorScheme } from "models/useColorScheme";

export const ProgressPanel = observer(() => {
  const { isError, isLoading, isReloading } = useDetectedInstallationStore();
  const { tone, colorScheme } = useColorScheme();

  if (isError) return null;
  if (!isLoading && !isReloading) return null;

  return (
    <>
      <Progress size="sm" isIndeterminate colorScheme={colorScheme} bg={tone(100)} borderRadius="lg" mb={1} />
      <Text display="block" textAlign="center" color={tone(800)} fontSize="sm" mb={6}>
        Processing
      </Text>
    </>
  );
});
