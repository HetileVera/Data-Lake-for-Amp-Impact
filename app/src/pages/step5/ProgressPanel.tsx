import React from "react";
import { Box } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useInstallation } from "AppContext";

import { OperationProgressPanel } from "pages/step5/OperationProgressPanel";

export const ProgressPanel = observer(() => {
  const installation = useInstallation();
  const deployment = installation.deploymentOperations;

  return (
    <Box p={3} pb={4}>
      {deployment.operations.map((operation, index) => (
        <OperationProgressPanel key={index} operation={operation} />
      ))}
      <Box mt="60px">
        <OperationProgressPanel operation={deployment} message="Overall Progress" />
      </Box>
    </Box>
  );
});
