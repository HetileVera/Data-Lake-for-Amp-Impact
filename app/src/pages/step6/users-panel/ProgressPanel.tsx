import React from "react";
import { Progress } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useInstallation } from "AppContext";
import { useColorScheme } from "models/useColorScheme";
import { useUsersStore } from "models/useUsersStore";

export const ProgressPanel = observer(() => {
  const installation = useInstallation();
  const { isError, isLoading, isReloading } = useUsersStore(installation);
  const { tone, colorScheme } = useColorScheme();

  if (isError) return null;
  if (!isLoading && !isReloading) return null;

  return <Progress size="sm" isIndeterminate colorScheme={colorScheme} bg={tone(100)} borderRadius="lg" mb={4} mt={0} />;
});
