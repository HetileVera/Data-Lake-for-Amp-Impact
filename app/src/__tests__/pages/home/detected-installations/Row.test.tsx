import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";

import { render, CustomChakraProvider } from "test-utils";

import { StoreProvider } from "AppContext";

import { ProgressPanel } from "pages/home/detected-installations/ProgressPanel";

import * as detectedInstallationStore from "models/useDetectedInstallationStore";
import { Row } from "pages/home/detected-installations/Row";
import { DetectedInstallation } from "models/helpers/DetectedInstallation";

const mockOnSelected = jest.fn();

describe("DetectedInstallations -> Row component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    const item = DetectedInstallation.create({
      id: "1",
      webUrlOrigin: "",
      reachable: true,
      region: "us-east-1",
      installationJson: {},
    });

    const tree = create(
      <CustomChakraProvider>
        <StoreProvider>
          <Row item={item} selected={false} onSelected={mockOnSelected} />
        </StoreProvider>
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly", () => {
    const item = DetectedInstallation.create({
      id: "1",
      webUrlOrigin: "",
      reachable: true,
      region: "us-east-1",
      installationJson: {},
    });

    render(
      <StoreProvider>
        <Row item={item} selected={false} onSelected={mockOnSelected} />
      </StoreProvider>
    );

    expect(screen.queryByText(/connection name/i)).toBeInTheDocument();
  });
});
