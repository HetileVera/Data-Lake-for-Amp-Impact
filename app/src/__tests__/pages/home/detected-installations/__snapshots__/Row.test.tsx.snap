// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetectedInstallations -> Row component snapshot 1`] = `
Array [
  <div
    className="css-yuhjaa"
    onClick={[Function]}
  >
    <svg
      className="chakra-icon css-1mu55r6"
      fill="currentColor"
      focusable={false}
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z"
      />
    </svg>
    <div
      className="css-1tw0ka0"
    >
      <div
        className="css-gg4vpm"
      >
        <p
          className="chakra-text css-2ygcmq"
        >
          Data Lake - 
          1
        </p>
        <p
          className="chakra-text css-dw5ttn"
        >
          US East (Northern Virginia) Region
        </p>
      </div>
      <div
        className="css-dw5ttn"
      >
        <p
          className="chakra-text css-5qx3eq"
        >
          â€¢
        </p>
        by 
      </div>
      <div
        className="css-dw5ttn"
      >
        Connection name: 
      </div>
    </div>
  </div>,
  <span />,
]
`;
